doc/src/manual/performance-tips.md

```diff
@@ -481,6 +481,64 @@ c = (b + 1.0f0)::Complex{T}
 does not hinder performance (but does not help either) since the compiler can determine the type of `c`
 at the time `k` is compiled.
 
+### Be aware of when Julia avoids specializing
+
+As a heuristic, Julia avoids automatically specializing on argument type parameters in three
+specific cases: `Type`, `Function`, and `Vararg`. Julia will always specialize when the argument is
+used within the method, but not if the argument is just passed through to another function. This
+usually has no performance impact at runtime and
+[improves compiler performance](@ref compiler-efficiency-issues). If you find it does have a
+performance impact at runtime in your case, you can trigger specialization by adding a type
+parameter to the method declaration. Here are some examples:
+
+This will not specialize:
+
+```julia
+function f_type(t)  # or t::Type
+    x = ones(t, 10)
+    return sum(map(sin, x))
+end
+```
+
+but this will:
+
+```julia
+function g_type(t::Type{T}) where T
+    x = ones(T, 10)
+    return sum(map(sin, x))
+end
+```
+
+These will not specialize:
+
+```julia
+f_func(f, num) = ntuple(f, div(num, 2))
+g_func(g::Function, num) = ntuple(g, div(num, 2))
+```
+
+but this will:
+
+```julia
+h_func(h::H, num) where {H} = ntuple(h, div(num, 2))
+```
+
+This will not specialize:
+
+```julia
+f_vararg(x::Int...) = tuple(x...)
+```
+
+but this will:
+
+```julia
+g_vararg(x::Vararg{Int, N}) where {N} = tuple(x...)
+```
+
+Note that [`@code_typed`](@ref) and friends will always show you specialized code, even if Julia
+would not normally specialize that method call. You need to check the
+[method internals](@ref ast-lowered-method) if you want to see whether specializations are generated
+when argument types are changed.
```

Comment:
Maybe mention that what `Test.@inferred` reports respects how Julia actually specialize the method call? (if that's correct)

https://discourse.julialang.org/t/inconsistent-code-warntype-and-inferred/24998
https://github.com/tkf/Transducers.jl/issues/14

Reply:
I worry about implying that the difference is simply about type inference. A method can have the same inferred return type but completely different specialization and runtime characteristics.

Reply:
^ Is that itself a comment worth adding?

Reply:
I don't think so. Lots of things have the same inferred return type to lots of other things.

Reply:
> I worry about implying that the difference is simply about type inference.

It makes sense.  But it also would be nice if there is an instruction to check if your code hits this case.  Maybe add an example REPL session invoking `(@which f(...)).specializations` etc. to demonstrate how to spot this issue?